### DISCUSSION TITLE
How to get peercert in hooks

### DISCUSSION THREAD

**Discussion by jcosentino11** [emqx/emqx#10644] [Q&A] [answered]
Hello!

Our plugin uses on_client_connect & authN/Z hooks, and in particular makes use of the `peercert` in the ConnInfo arg.  I noticed when upgrading emqx to 5.x that `peercert` is undefined, even with `verify_peer` set in ssl listener config.

If I'm understanding correctly, it seems like `peercert` has been removed from connection info with https://github.com/emqx/emqx/pull/10243/files?  It mentions to pull this info from the socket instead, what would be the proper way to do that within a hook?

EDIT: same question as https://github.com/emqx/emqx/pull/10243/files#r1187724879

thanks in advance!

**Comment by zmstone** [comment, marked as answer]
Hi @jcosentino11 
It will be fixed in this PR: https://github.com/emqx/emqx/pull/10715
thank you for bring this to our attention.

**Comment by keynslug** [comment]
Hi @jcosentino11, thanks for bringing this up!

We brought back the former behavior with https://github.com/emqx/emqx/commit/b2ecbef0f167891cc55b220d6ce41facfb6bdd99, so with next release you should be able to use `peercert` in `ConnInfo` as before.

In the meantime you should be able to hook `client.authenticate` hook point, which is provided with [`ClientInfo`](https://github.com/emqx/emqx/blob/v5.0.25/apps/emqx/src/emqx_types.erl#L148-L165) structure when invoked [[1](https://github.com/emqx/emqx/blob/v5.0.25/apps/emqx/src/emqx_access_control.erl#L52)] [[2](https://github.com/emqx/emqx/blob/v5.0.25/apps/emqx/src/emqx_access_control.erl#L120)]. You could configure `mqtt.peer_cert_as_username = crt` (globally or per specific zone) and then get peer certificate verbatim in [`ClientInfo = #{username := PeerCert}`](https://github.com/emqx/emqx/blob/v5.0.25/apps/emqx/src/emqx_types.erl#L154). Also worth noting that irrespective of this configuration parameter, the `ClientInfo` structure will always contain certificate's CN and DN.
